cni/pkg/nodeagent/cni-watcher.go

```diff
@@ -0,0 +1,182 @@
+func (s *CniPluginServer) Start() error {
+	if s.sockAddress == "" {
+		return fmt.Errorf("no socket address provided")
+	}
+	log.Info("Start a listen server for CNI plugin events")
+	unixListener, err := pluginlistener.NewListener(s.sockAddress)
+	if err != nil {
+		return fmt.Errorf("failed to create CNI listener: %v", err)
+	}
+	go func() {
+		if err := s.cniListenServer.Serve(unixListener); network.IsUnexpectedListenerError(err) {
+			log.Errorf("Error running CNI listener server: %v", err)
+		}
+	}()
+
+	go func() {
+		<-s.ctx.Done()
+		if err := s.cniListenServer.Close(); err != nil {
+			log.Errorf("CNI listen server terminated with error: %v", err)
+		} else {
+			log.Debug("CNI listen server terminated")
+		}
+	}()
+
+	return nil
+}
+
+func (s *CniPluginServer) handleAddEvent(w http.ResponseWriter, req *http.Request) {
+	if req.Body == nil {
+		log.Error("empty request body")
+		http.Error(w, "empty request body", http.StatusBadRequest)
+		return
+	}
+	defer req.Body.Close()
+	data, err := io.ReadAll(req.Body)
+	if err != nil {
+		log.Errorf("Failed to read event report from cni plugin: %v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	msg, err := processAddEvent(data)
+	if err != nil {
+		log.Errorf("Failed to process CNI event payload: %v", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	if err := s.ReconcileCNIAddEvent(req.Context(), msg); err != nil {
+		log.Errorf("Failed to handle add event: %v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+func processAddEvent(body []byte) (CNIPluginAddEvent, error) {
+	var msg CNIPluginAddEvent
+	err := json.Unmarshal(body, &msg)
+	if err != nil {
+		log.Errorf("Failed to unmarshal CNI plugin event: %v", err)
+		return msg, err
+	}
+
+	log.Debugf("Deserialized CNI plugin event: %+v", msg)
+	return msg, nil
+}
+
+func (s *CniPluginServer) ReconcileCNIAddEvent(ctx context.Context, addCmd CNIPluginAddEvent) error {
+	log := log.WithLabels("cni-event", addCmd)
+
+	// The CNI node plugin should have already checked the pod with this
+	// exact same function before forwarding us the event, but we have to invoke the K8S client anyway,
+	// so to be safe we check it again here to make sure we get the same result
+	ambientPod, err := s.handlers.GetPodIfAmbient(addCmd.PodName, addCmd.PodNamespace)
+	if err != nil {
+		return err
+	}
+
+	if ambientPod != nil {
+		log.Debugf("Pod: %s in ns: %s is enabled for ambient, adding to mesh. ", addCmd.PodName, addCmd.PodNamespace)
+
+		var podIps []netip.Addr
+		for _, configuredPodIPs := range addCmd.IPs {
+			// net.ip is implicitly convertible to netip as slice
+			ip, _ := netip.AddrFromSlice(configuredPodIPs.Address.IP)
+			// We ignore the mask of the IPNet - it's fine if the IPNet defines
+			// a block grant of addresses, we just need one for checking routes.
+			podIps = append(podIps, ip)
```

Comment:
We cannot, because the K8S pod spec doesn't get it until after the CNI execution finishes, and we are in the middle of CNI execution at this point.

The CNI is what provides the pod IP to the K8S pod spec in the first place, when it is finally populated, so this is guaranteed to be both early and correct.

Reply:
good to know